kolibri/plugins/coach/assets/src/composables/useFetch.js

```diff
@@ -0,0 +1,199 @@
+
+/**
+ * A composable for managing fetch operations with optional methods for additional data fetching.
+ *
+ * @param {Object} options **Required** Configuration options for the fetch operation.
+ * @param {(...args) => Promise<any>} options.fetchMethod **Required** Function to fetch the
+ * initial data. Should return a Promise resolving to the fetched data.
+ *
+ * Example:
+ * ```js
+ * const fetchMethod = () => ContentNodeResource.fetchBookmarks({
+ *   params: { limit: 25, available: true, },
+ * })
+ *
+ *
+ * @param {string} [options.dataKey] The key in the response object where the main data is located.
+ * * You can use `lodash.get`-compatible keys to access nested objects (e.g., "data.user.details").
+ * * If `dataKey` is left `undefined`, the entire response object will be treated as the main data.
+ *
+ * Example:
+ * ```js
+ * // Suppose the response object looks like this:
+ * const response = {
+ *   status: "success",
+ *   payload: { id: 42, name: "Jane Doe" }
+ * };
+ *
+ * // By specifying `dataKey`, you tell the composable where to find the main data:
+ * const { data } = useFetch({ dataKey: "payload" });
+ * console.log(data); // Outputs: { id: 42, name: "Jane Doe" }
+ * ```
+ *
+ *
+ * @param {string} [options.moreKey] The key in the response object where the
+ * "more" object is located.
+ * * This is the object that will be passed as the first argument to the `fetchMoreMethod` param.
+ * * You can use `lodash.get`-compatible keys to access nested objects (e.g., "data.user.details").
+ *
+ * Example:
+ * ```js
+ * // Suppose the response object looks like this:
+ * const response = {
+ *   payload: [...],
+ *   more: { page: 2 }
+ * };
+ *
+ * // By specifying `moreKey`, you tell the composable where to find the more object
+ * // to fetch additional data:
+ * useFetch({ moreKey: "more" });
+ * ```
+ *
+ *
+ * @param {(more, ...args) => Promise<any>} [options.fetchMoreMethod] Function to fetch more data.
+ * * This function receives a "more" object as its first argument. This "more" object is specified
+ *   by the `moreKey` param.
+ * * Should return a Promise resolving to the additional data.
+ * * FetchMore just works if the fetched data is an array
+ *
+ * Example:
+ *
+ * ```js
+ * const fetchMoreMethod = (more) => ContentNodeResource.fetchBookmarks({
+ *  params: more
+ * })
+ * ```
+ *
+ *
+ * @param {Object.<string, string>} [options.additionalDataKeys] An object that maps additional
+ * data keys for the response object.
+ * * In the `{ key: value }` pair:
+ *   - The `key` will be used as the property name in the returned `additionalData` object.
+ *   - The `value` specifies the key in the response object from which the data will be retrieved.
+ *
+ * Example:
+ * ```js
+ * const additionalDataKeys = {
+ *   userId: "user_id", // The `userId` property in `additionalData` will map to `response.user_id`
+ *   userName: "name" // The `userName` property in `additionalData` will map to `response.name`
```

Comment:
No. The intention is to separate the actual "data" (i.e. if we are fetching bookmarks, have in the `data` ref the actual array of bookmarks) from the rest of the response (i.e. the response of the bookmarks fetch can contain additional data like the `count` that tells the total amount of bookmarks, not just the length of the responded array), in part to have a cleaner api, and also to improve the process of "fetching more data".

For example when fetching bookmarks we can  get a response like:

```javascript
{
  results: [ ... ], // lets say limited to 25 bookmarks per fetch
  count: 432.
  more: {...} // this more object will contain the query needed to fetch more data or to fetch the "next page"
}
```

Then we want `data` to be the bookmarks array, no the full response, so when we fetch more bookmarks using the `fetchMore()` method, we already have the actual array in the `data` ref, and we know how to merge them. Thats why in this case `dataKey = "results"`.

But, apart from just the array of bookmarks, in the UI we also show the total number of bookmarks the user has, which is different from the data.length value, because the response is paginated. So to know the total number of bookmarks we need can use the `count` key that comes in the same response. Apart from that we can imagine that the endpoint also returns the weight of the bookmarks in a key `weight` or any other metadata apart from the actual array of bookmarks. As we want to show this "additional data" (additional to the main data which is the bookmarks array) we need to query them and store them in a ref. But the problem is that this data is dynamic, so its difficult to create a `count` ref, a `weight` ref, etc... to store all this additional data. For this we have this `additionalDataKeys` object, that will tell to the useFetch something like "hey apart from the bookmarks array I also want the "count" data that is stored in "response.someNestedObject.count"". So we write a `additionalDataKeys` key-value pair object like `{ count: "someNestedObject.count" }` to tell that we want the "response.someNestedObject.count" data stored in the "count" key in the "additionalData" object that we return from the useFetch.

So after we fetch bookmarks we can use this count.
```
const { additionalData, .... } = useFetch(...)

console.log(additionalData.count) // 432.

Reply:
My sense here is that the `additionalDataKeys` is a premature abstraction at this point - we are only using it twice in this PR, and for seemingly very different purposes, one is to return a specific piece of response data - the `count`, the other is to return the entirety of the response data.

This suggests to me that trying to abstract away everything behind options configuration might be a wrong turn, and we should limit that to the truly repeated functionality - loading state, loading state for more items for pagination, and maybe fetching more.

Reply:
The best alternative I can think is just returning the whole response object apart from the `data` ref?

Reply:
The best alternative I can think of is to return the whole response object apart from the `data` ref, how does that sounds?

Reply:
See my suggestion for `fetchTree` specifically - I think that does away with the need for a lot of the abstraction here, `count` we can just always return and just have it be `null` in the case when the pagination is cursor based, and does not provide a count.

Reply:
Updated :). Thank you @rtibbles :open_hands: